### ISSUE TITLE
FR: Make plum work with ipython's autoreload

### ISSUE THREAD

**Issue by PhilipVinc** [wesselb/plum#38]
The two don't work well together. 
I think it's because autoreload changes the classes of objects, but for some reason forgets to update the Types in plum's dispatch tables.

Should be investigated...

**Comment by wesselb** [comment]
Also tagging @seeM here.

This issue is unfortunately a tricky one to fix. Just to document what's going wrong, here's a minimal example.

Consider the minimal module

```python
from __future__ import annotations


class A:
    def test(self) -> A:
        # Hello
        return self
```

We now have the following behaviour:

```python
In [1]: from core import A

In [2]: from typing import get_type_hints

In [3]: get_type_hints(A.test)["return"] == A
Out[3]: True

In [4]: # Make changes and trigger autoreload

In [5]: get_type_hints(A.test)["return"] == A
Out[5]: False
```

(This is on Python 3.9.)

I think it's necessary to carefully pin down how `get_type_hints` interacts with `autoreload`, especially with regards to forwards references.

If this particular case can be fixed, I think good support for `autoreload` in the case of `from __future__ import annotations` should be possible.

**Comment by PhilipVinc** [comment]
I had investigated the issue a bit, and eventually realized that the best course of action would be for someone to contact the developers of iPython to ask them to add an hook to auto reload so that when he reloads a class/files plugins can ask to be notified about it.

**Comment by wesselb** [comment]
@PhilipVinc That's interesting. I suppose hooking into autoreload would be the way go about this: upon every autoreload, clear all cache.

I'm wondering, is the above example expected behaviour? Should `get_type_hints(A.test)["return"] == A` after autoreloading? The locals in an iPython sessions should be updated with every autoreload, right?

Perhaps not relevant to this issue, but I'm tempted to strip out all current functionality to handle forward references from the package and just rely on `typing.get_type_hints` to handle forward references. What are your thoughts on this?

**Comment by wesselb** [comment]
Experimental support thanks to #44 by @PhilipVinc is included in the latest release `v1.6`. :)
